From patchwork Tue Jan 12 09:29:42 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Kunkun Jiang <jiangkunkun@huawei.com>
X-Patchwork-Id: 12012889
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-16.8 required=3.0 tests=BAYES_00,
	HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,INCLUDES_PATCH,
	MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,URIBL_BLOCKED,USER_AGENT_GIT
	autolearn=ham autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 4F296C433DB
	for <kvm@archiver.kernel.org>; Tue, 12 Jan 2021 09:33:05 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 25ED722D58
	for <kvm@archiver.kernel.org>; Tue, 12 Jan 2021 09:33:05 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S2392594AbhALJcs (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 12 Jan 2021 04:32:48 -0500
Received: from szxga04-in.huawei.com ([45.249.212.190]:10651 "EHLO
        szxga04-in.huawei.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S2390802AbhALJcs (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 12 Jan 2021 04:32:48 -0500
Received: from DGGEMS406-HUB.china.huawei.com (unknown [172.30.72.58])
        by szxga04-in.huawei.com (SkyGuard) with ESMTP id 4DFQLK10Nyz15rbR;
        Tue, 12 Jan 2021 17:31:05 +0800 (CST)
Received: from DESKTOP-6NKE0BC.china.huawei.com (10.174.185.210) by
 DGGEMS406-HUB.china.huawei.com (10.3.19.206) with Microsoft SMTP Server id
 14.3.498.0; Tue, 12 Jan 2021 17:31:58 +0800
From: Kunkun Jiang <jiangkunkun@huawei.com>
To: Paolo Bonzini <pbonzini@redhat.com>, <kvm@vger.kernel.org>,
        <linux-kernel@vger.kernel.org>
CC: <wanghaibin.wang@huawei.com>, Keqian Zhu <zhukeqian1@huawei.com>
Subject: [PATCH] kvm: Fixes lack of KVM_CAP_MANUAL_DIRTY_LOG_PROTECT2 enabled
 check
Date: Tue, 12 Jan 2021 17:29:42 +0800
Message-ID: <20210112092942.2310-1-jiangkunkun@huawei.com>
X-Mailer: git-send-email 2.26.2.windows.1
MIME-Version: 1.0
X-Originating-IP: [10.174.185.210]
X-CFilter-Loop: Reflected
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

The KVM_CLEAR_DIRTY_LOG ioctl lacks the check whether the capability
KVM_CAP_MANUAL_DIRTY_LOG_PROTECT is enabled or not. This may cause
some problems if userspace calls the KVM_CLEAR_DIRTY_LOG ioctl, but
dose't enable this capability. So we'd better to add it.

Fixes: 2a31b9db15353 ("kvm: introduce manual dirty log reprotect")
Signed-off-by: Kunkun Jiang <jiangkunkun@huawei.com>
---
 virt/kvm/kvm_main.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/virt/kvm/kvm_main.c b/virt/kvm/kvm_main.c
index fa9e3614d30e..8f5633d8a0e8 100644
--- a/virt/kvm/kvm_main.c
+++ b/virt/kvm/kvm_main.c
@@ -1602,6 +1602,9 @@ static int kvm_clear_dirty_log_protect(struct kvm *kvm,
 	unsigned long *dirty_bitmap_buffer;
 	bool flush;
 
+	if (!kvm->manual_dirty_log_protect)
+		return -EPERM;
+
 	/* Dirty ring tracking is exclusive to dirty log tracking */
 	if (kvm->dirty_ring_size)
 		return -ENXIO;
